Pend cleanup

RussTedrake opened the pull request:
This are the first few commits in a series that will (hopefully) dramatically clean up and flesh out the pendulum example directory.  I hope this is a manageable enough chunk, and have tried to bite off separate commits for easier reviewing (the last commit sprawled on me a little bit).

fyi @jadecastro -- this also resolves a bug/missing feature in the linearization.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6940)
<!-- Reviewable:end -->

RussTedrake commented:
+@jwnimmer-tri for feature review, please.  

---

Review status: 0 of 13 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-:-Ksnjq79skplFwrddaE1:b-x0ad30)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 4 of 17 files at r1.
Review status: 3 of 20 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/systems/primitives/linear_system.cc, line 102 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspWjwxV5I9jpAYLkuw:-KspWjwxV5I9jpAYLkux:b-lexnik) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5c761b320f8ed0f5f38a18ba2df0e20a9c99dc/drake/systems/primitives/linear_system.cc#L102)):*
> ```cpp
>   }
> 
>   auto autodiff_args = math::initializeAutoDiffTuple(x0, u0);
> ```

FYI I found this difficult to follow.  I think using a `std::tie` would be more clear.  Off the top of my head, that would look like:
```
AutoDiffXd autodiff_x;
AutoDiffXd autodiff_u;
std::tie(autodiff_x, autodiff_u) = math::initializeAutoDiffTuple(x0, u0);
autodiff_context->get_mutable_continuous_state_vector()->SetFromVector(
    autodiff_x);
if (num_inputs > 0) {
  autodiff_context->FixInputPort(0, autodiff_u);
}
```

---

*[drake/systems/primitives/linear_system.cc, line 116 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspWVDFkK4cmRztxm6v:-KspWVDFkK4cmRztxm6w:b5kw3qa) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5c761b320f8ed0f5f38a18ba2df0e20a9c99dc/drake/systems/primitives/linear_system.cc#L116)):*
> ```cpp
>   // Set derivatives of all parameters in the context to zero (but with the
>   // correct size).
>   int num_gradients = num_states + num_inputs;
> ```

Would this be clearer as `int num_gradients = std::get<0>(autodiff_args).derivs().size();` (or `autodiff_x.derivs.size()` with the new local variable suggestion, above), since I think the point is to match the other derivative vectors?

---

*[drake/systems/primitives/linear_system.cc, line 124 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Ksp_F1K-taIy5ipLVaH:-Ksp_F1LXm7BpMfxLzue:b-buo54q) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L124)):*
> ```cpp
>     for (int j = 0; j < params.size(); j++) {
>       auto& derivs = params(j).derivatives();
>       if (derivs.size() == 0) {
> ```

I can't figure out when this would end up not true -- perhaps if some system's AllocateParameters did a special-case override for AutoDiffXd that structured the derivs already somehow?  Probably a comment here is in order explaining why we have this guard.

Is there unit test coverage of the else condition (that size was non-zero, and was left unchanged)?

---

*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-KspYWwcJ_dKaEK0Q0Xu:b-gpb2wh) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
> ```cpp
>       }
>     }
>   }
> ```

I am struggling to understand why literal constants (with zero for all derivatives) need to have explicitly sized derivative vectors.  I can see why `autodiff_x` needs to be initialized... you need the `∂x[i]/∂x[i] == 1` to be populated.  But if my plant just did like a `y = x * 2.0;` for its output and the 2.0 was a C++ literal and not a Parameter, stuff is still supposed to work!  My expectation is that derivative-less doubles promote into `AutoDiffXd` with empty derivs, which AutoDiff treats as if the derivatives were zero, so I can't see why this resizing would be necessary.

I haven't finished reading the PR yet -- is there some calling code in the pendulum that you can highlight for me that depends on this resizing?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 11 of 17 files at r1, 7 of 7 files at r2.
Review status: 18 of 20 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[drake/examples/pendulum/lqr_simulation.cc, line 99 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspeW-ofBs0QgzxMV6Z:-KspeW-ofBs0QgzxMV6_:b7hb162) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/lqr_simulation.cc#L99)):*
> ```cpp
> }  // namespace drake
> 
> int main() { return drake::examples::pendulum::do_main(); }
> ```

This needs `gflags::ParseCommandLineFlags(&argc, &argv, true);` in order for the command-line parameters to take effect.

---

*[drake/examples/pendulum/passive_simulation.cc, line 71 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspeaWYAMFQccbY1M5E:-KspeaWZEVfqGjuCDtwC:b7hb162) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/passive_simulation.cc#L71)):*
> ```cpp
> }  // namespace drake
> 
> int main() { return drake::examples::pendulum::do_main(); }
> ```

This needs `gflags::ParseCommandLineFlags(&argc, &argv, true);` in order for the command-line parameters to take effect.

---

*[drake/examples/pendulum/pendulum_params.named_vector, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspdXXzI8T785UuZq7n:-KspdXY-Gt1ecRaPtCkQ:b6iouz0) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/pendulum_params.named_vector#L1)):*
> ```
> # The defaults in this file approximate a 2010 Toyota Prius.
> ```

Unlikely ;)

---

*[drake/examples/pendulum/pendulum_plant.cc, line 18 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspbjNWp_UUoABqupS6:-KspbjNXSYGPU5BjlD47:b97d7rz) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/pendulum_plant.cc#L18)):*
> ```cpp
>                                 &PendulumPlant::CopyStateOut);
>   this->DeclareContinuousState(PendulumState<T>(), 1 /* num_q */, 1 /* num_v */,
>                                0 /* num_z */);
> ```

FYI We probably could (and should) promote the q/v/z count into the vector class's code-gen, so that they can be omitted here (or at worst, cited as named constants).  (Not for this PR.)

---

*[drake/examples/pendulum/pendulum_plant.cc, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Kspc4QD_Dp7Kh2aRMmw:-Kspc4QD_Dp7Kh2aRMmx:bdbx4rv) ([raw file](https://github.com/robotlocomotion/drake/blob/0ec3eba238258559c341a031564e7d2d0c2928e6/drake/examples/pendulum/pendulum_plant.cc#L20)):*
> ```cpp
>       PendulumStateVector<T>(),
>       1 /* num_q */, 1 /* num_v */, 0 /* num_z */);
>   static_assert(PendulumStateVectorIndices::kNumCoordinates == 1 + 1, "");
> ```

Why lose this check?

---

*[drake/examples/pendulum/pendulum_plant.cc, line 47 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspcpmIRIi2x_mcG2pa:-KspcpmIRIi2x_mcG2pb:bw3gsre) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/pendulum_plant.cc#L47)):*
> ```cpp
> template <typename T>
> T PendulumPlant<T>::CalcTotalEnergy(const systems::Context<T>& context) const {
>   using std::pow;
> ```

Needs `#include <cmath>`.

---

*[drake/examples/pendulum/pendulum_plant.cc, line 51 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspczyDImDIWqBggZjm:-KspczyDImDIWqBggZjn:b-fgfygh) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/pendulum_plant.cc#L51)):*
> ```cpp
>   const PendulumParams<T>& params =
>       this->template GetNumericParameter<PendulumParams>(context, 0);
>   // Kinetic energy = 1/2 m l² θ̇².
> ```

Both github and reviewable are rendering the unicode dot in thetadot atop the squared, not atop the theta (I had to zoom _way_ in to find it).  Is there some better unicode to use?

---

*[drake/examples/pendulum/pendulum_plant.cc, line 58 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspfK6plvvRabHh8n65:-KspfK6plvvRabHh8n66:b-308t8z) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/pendulum_plant.cc#L58)):*
> ```cpp
>       -params.mass() * params.gravity() * params.length() * cos(state.theta());
>   return kinetic_energy + potential_energy;
> }
> ```

This is a beautiful implementation, but probably should have a (tiny) unit test to check it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 17 files at r1.
Review status: 19 of 20 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[drake/examples/pendulum/energy_shaping_simulation.cc, line 42 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KsphNQZGLkYjX-rUdOW:-KsphNQZGLkYjX-rUdOX:b-7r7ah7) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/energy_shaping_simulation.cc#L42)):*
> ```cpp
>     const PendulumState<T>* const state = dynamic_cast<const PendulumState<T>*>(
>         this->EvalVectorInput(context, 0));
>     DRAKE_ASSERT(state != nullptr);
> ```

FYI Wow, it sucks that `const auto& state = *(this->EvalVectorInput<PendulumState>(context, 0));` doesn't do what we want.  I wonder why it is documented to return nullptr rather than throw or abort?

In any case, consider `const auto* state = this->EvalVectorInput<PendulumState>(context, 0);` which seems like it should work, and pushes the cast gunk out of our way here.

---

*[drake/examples/pendulum/energy_shaping_simulation.cc, line 52 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Kspi9n2z7vY6x-3QWxx:-Kspi9n2z7vY6x-3QWxy:bspsjwr) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/energy_shaping_simulation.cc#L52)):*
> ```cpp
>     // http://underactuated.csail.mit.edu/underactuated.html?chapter=3
>     using std::pow;
>     const T Etilde = .5 * params.mass() * pow(params.length(), 2) *
> ```

FYI If you do `const T Etilde = // BR` to force a linebreak, you'll get ~14 more characters to work with, and I suspect it would linewrap more readably.

---

*[drake/examples/pendulum/energy_shaping_simulation.cc, line 53 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Kspi3aaqip6wbcUIjVD:-Kspi3abrVSCKSARPff4:b-k9jcy0) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/energy_shaping_simulation.cc#L53)):*
> ```cpp
>     using std::pow;
>     const T Etilde = .5 * params.mass() * pow(params.length(), 2) *
>                          state->thetadot() * state->thetadot() -
> ```

Why sometimes `pow(foo, 2)` and sometimes `foo * foo`?

---

*[drake/examples/pendulum/energy_shaping_simulation.cc, line 73 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Ksph-falbBxxPQmP9C_:-Ksph-falbBxxPQmP9Ca:b-5f4h08) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/energy_shaping_simulation.cc#L73)):*
> ```cpp
>   pendulum->set_name("pendulum");
>   auto controller = builder.AddSystem<PendulumEnergyShapingController>(
>       PendulumParams<double>());
> ```

FYI Should there be a TODO that the controller and plant somehow share or correspond their params?

---

*[drake/examples/pendulum/energy_shaping_simulation.cc, line 106 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspgQns2TYF5OaVopcp:-KspgQns2TYF5OaVopcq:b311mxu) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/energy_shaping_simulation.cc#L106)):*
> ```cpp
>   // towards the desired energy level.
>   DRAKE_DEMAND(std::abs(initial_energy - desired_energy) >
>                2.0 * std::abs(final_energy - desired_energy));
> ```

FYI This took me a little while to parse.  My final answer was "the discrepancy between our energy and the desired energy has at least halved during the simulation", which might be more clearly written as:
```
DRAKE_DEMAND(
    std::abs(final_energy - desired_energy) <
    (0.5 * std::abs(initial_energy - desired_energy)));
```

---

*[drake/examples/pendulum/energy_shaping_simulation.cc, line 116 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Kspfe_GGH5RqygImh8V:-Kspfe_H6wwPbvajgmqh:b7hb162) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/energy_shaping_simulation.cc#L116)):*
> ```cpp
> }  // namespace drake
> 
> int main() { return drake::examples::pendulum::do_main(); }
> ```

This needs `gflags::ParseCommandLineFlags(&argc, &argv, true);` in order for the command-line parameters to take effect.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 19 of 20 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[drake/systems/primitives/BUILD, line 324 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Kspj8QWuHlIWPqHxnFb:-Kspj8QXB3PINEs9V0lq:b-phtvt5) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/BUILD#L324)):*
> ```Python
>         ":linear_system",
>         "//drake/common:eigen_matrix_compare",
>         "//drake/examples/pendulum:pendulum_plant",
> ```

CI is unhappy because `primitives` still has `CMakeLists.txt` but it doesn't mention pendulum (nor car it, because there's no CMake there).  I would say just de-list this unit test from CMake.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 19 of 20 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-KspifbQ1YnHogCYBEJn:bh0v0xn) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I am struggling to understand why literal constants (with zero for all derivatives) need to have explicitly sized derivative vectors.  I can see why `autodiff_x` needs to be initialized... you need the `∂x[i]/∂x[i] == 1` to be populated.  But if my plant just did like a `y = x * 2.0;` for its output and the 2.0 was a C++ literal and not a Parameter, stuff is still supposed to work!  My expectation is that derivative-less doubles promote into `AutoDiffXd` with empty derivs, which AutoDiff treats as if the derivatives were zero, so I can't see why this resizing would be necessary.

I haven't finished reading the PR yet -- is there some calling code in the pendulum that you can highlight for me that depends on this resizing?
</blockquote></details>

The unit test I wrote for linear_system_test fails without it.  `y = x * 2.0` works fine, but `z = x * y` with `x` and `y` both autodiff but only x has a derivative vector with non-zero derivatives size results in a "you multiplied matrices of mismatched sizes" assertion in debug mode, and simply zero gradients for `z` in release mode.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-Kspjao01xRK1dzczI2H:bw3c86u) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

The unit test I wrote for linear_system_test fails without it.  `y = x * 2.0` works fine, but `z = x * y` with `x` and `y` both autodiff but only x has a derivative vector with non-zero derivatives size results in a "you multiplied matrices of mismatched sizes" assertion in debug mode, and simply zero gradients for `z` in release mode.
</blockquote></details>

So it sounds like if I happened to write `y = x * T(2.0);` it would similarly warn / fail?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 19 of 20 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-Kspk51u8WeQr9Fl0If9:b-erht6b) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

So it sounds like if I happened to write `y = x * T(2.0);` it would similarly warn / fail?
</blockquote></details>

Yes, I believe so (because that is effectively what was happening here... the params were set by context.SetTimeStateAndParameters from a Context<double>

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r3.
Review status: 20 of 21 files reviewed at latest revision, 19 unresolved discussions.

---

*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-KspkpoewvLxWATvItLB:bvr1yup) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Yes, I believe so (because that is effectively what was happening here... the params were set by context.SetTimeStateAndParameters from a Context<double>
</blockquote></details>

That is astonishing and suggests to me that AutoDiff is a minefield that will be astonishingly difficult to use.  Even within only the framework, we already have:
```console
find . -name test -prune -o -type f -print0 | xargs -0 grep -n -e '^[^/]*[^R]T('
./vector_base.h:90:      SetAtIndex(i, T(0));
./vector_base.h:143:    return PlusEqScaled(T(1), rhs);
./vector_base.h:148:    return PlusEqScaled(T(-1), rhs);
./witness_function.h:126:  decltype(T() < T()) should_trigger(const T& w0, const T& wf) const {
./witness_function.h:132:        return (T(0) > T(0));
./basic_vector.h:147:    data->SetAtIndex(index++, T(constructor_arg));
./basic_vector.h:155:    data->SetAtIndex(index++, T(constructor_arg));
./system.h:1082:          our_vec->SetAtIndex(j, T(other_vec->GetAtIndex(j)));
./system.h:1347:    return T(0);
./system.h:1357:    return T(0);
./system.h:1370:    return T(0);
./system.h:1387:    return T(0);
./context.h:420:    set_time(T(source.get_time()));
```

If you grep the whole tree, we're north of 100 `T(double)` uses already.  It sounds like we will need coding standard advice for this.

Even taking that as given, if indeed every `T` in a `Context<AutoDiffXd>` needs to have the same `derivs.size()`, isn't that something that the Context should take care of for us (either automatically, or with an opt-in "fix all the things" call)?  Why should this code live in `linear_system` only?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-Kspnz3YhVpgeSIKzWlx:bto75kb) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
> AutoDiff is a minefield that will be astonishingly difficult to use.
> ./system.h:1347:    return T(0);

... and that's not even the half of it.  If this was just `return 0.0` it would implicitly convert to `AutoDiffXd{0.0}`, and then we have _explosions_ if we ever do arithmetic with that.  So there's probably many more sites than my `grep` found.  Anytime you promote a `double` to a `T`, rather than math-operate a `double` onto a pre-existing `T`, _explosions_.

It seems to me like `Xd` is broken and cannot be used, and we should move to `8d` or `32d` ASAP, since we want to do that anyway.

> Why should this code live in linear_system only?

For this PR, it could be a TODO(issue) to generalize the code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 20 of 21 files reviewed at latest revision, 19 unresolved discussions.

---

*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-Kspp_OTiIU0AzmSPM0J:b6a2vlh) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

> AutoDiff is a minefield that will be astonishingly difficult to use.
> ./system.h:1347:    return T(0);

... and that's not even the half of it.  If this was just `return 0.0` it would implicitly convert to `AutoDiffXd{0.0}`, and then we have _explosions_ if we ever do arithmetic with that.  So there's probably many more sites than my `grep` found.  Anytime you promote a `double` to a `T`, rather than math-operate a `double` onto a pre-existing `T`, _explosions_.

It seems to me like `Xd` is broken and cannot be used, and we should move to `8d` or `32d` ASAP, since we want to do that anyway.

> Why should this code live in linear_system only?

For this PR, it could be a TODO(issue) to generalize the code.
</blockquote></details>

I will run a few tests to verify in the morning.  What I know for sure is that after I set parameters in the pendulum, the lqr test broke, and that after writing the unit test here, it failed with the linear system matrices printed as all zeros until I inserted this code to fix the test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-KspqiaoHsx-dmR60ske:bhwcre0) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

I will run a few tests to verify in the morning.  What I know for sure is that after I set parameters in the pendulum, the lqr test broke, and that after writing the unit test here, it failed with the linear system matrices printed as all zeros until I inserted this code to fix the test.
</blockquote></details>

Reading Eigen code, it looks like `AutoDiffScalar` tries to `internal::make_coherent` to promote empty derivatives up to match the other operand in many cases, so perhaps its not as bad as I thought.  Possibly there is just a case missing for some operand we're using, or possibly one of _our_ operators is missing the coherency logic.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 20 of 21 files reviewed at latest revision, 19 unresolved discussions.

---

*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-Ksps7bF9AsKKGEHmOUb:b-9e568w) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Reading Eigen code, it looks like `AutoDiffScalar` tries to `internal::make_coherent` to promote empty derivatives up to match the other operand in many cases, so perhaps its not as bad as I thought.  Possibly there is just a case missing for some operand we're using, or possibly one of _our_ operators is missing the coherency logic.
</blockquote></details>

great.  then I will step through (tomorrow) and find out where things go bad.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 20 of 21 files reviewed at latest revision, 19 unresolved discussions.

---

*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-Kspx3bIIcS0C7HVwdn5:bhv3bim) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

great.  then I will step through (tomorrow) and find out where things go bad.
</blockquote></details>

ok.  i played a bit afterall ;-)  i left params to have empty derivatives, state has initialized derivatives, and get
```
T test = params.mass() * state.theta();
```
is ok.
```
T test = params.mass() * params.length() * state.theta();
```
blows up (SIGABRT in debug mode).
```
T test = params.mass() * params.length();
```
blows up (SIGABRT in debug mode).

so I guess it has trouble making things coherent if you are multiplying two empty derivative vectors together.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-Ksq2acCvXb2m2bJvwei:b9xpfw8) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
Great.

The problem is in
```c++
    operator*(const AutoDiffScalar<OtherDerType>& other) const
    {
      internal::make_coherent(m_derivatives, other.derivatives());
      return MakeAutoDiffScalar(
        m_value * other.value(),
        (m_derivatives * other.value()) + (other.derivatives() * m_value));
    }
```
on the `(m_derivatives * other.value()) + (other.derivatives() * m_value));` part, with the failing assertion being `Assertion 'aLhs.rows() == aRhs.rows() && aLhs.cols() == aRhs.cols()' failed`.

The type of that expression is
```c++
Eigen::CwiseBinaryOp<
  Eigen::internal::scalar_sum_op<double, double>,
  const Eigen::CwiseBinaryOp<
    Eigen::internal::scalar_product_op<double, double>,
    const Eigen::CwiseBinaryOp<
      Eigen::internal::scalar_sum_op<double, double>,
      const Eigen::CwiseBinaryOp<
        Eigen::internal::scalar_product_op<double, double>,
        const Eigen::VectorXd,
        const Eigen::CwiseNullaryOp<
          Eigen::internal::scalar_constant_op<double>,
          const Eigen::VectorXd
        >
      >,
      const Eigen::CwiseBinaryOp<
        Eigen::internal::scalar_product_op<double, double>,
        const Eigen::VectorXd,
        const Eigen::CwiseNullaryOp<
          Eigen::internal::scalar_constant_op<double>,
          const Eigen::VectorXd
        >
      >
    >,
    const Eigen::CwiseNullaryOp<
      Eigen::internal::scalar_constant_op<double>,
      const Eigen::VectorXd
    >
  >,
  const Eigen::CwiseBinaryOp<
    Eigen::internal::scalar_product_op<double, double>,
    const Eigen::VectorXd,
    const Eigen::CwiseNullaryOp<
      Eigen::internal::scalar_constant_op<double>,
      const Eigen::VectorXd
    >
  >
>::CwiseBinaryOp(const Lhs &, const Rhs &, const BinaryOp &)
```

The coherency promotion signature is:
```c++
struct make_coherent_impl<Matrix<A_Scalar, A_Rows, A_Cols, A_Options, A_MaxRows, A_MaxCols>,
                             Matrix<B_Scalar, B_Rows, B_Cols, B_Options, B_MaxRows, B_MaxCols> > {
```

My guess is that the Cwise intermediate results are not matching the `Matrix` specialization of `make_coherent_impl`, but rather hitting the no-op default implementation of it.

If you force any of the intermediate multiplies back into a `T`, instead of a promise, I think it should stop failing, e.g.,
```c++
T test = T{params.mass() * params.length()} * state.theta();
```
This might be the best work-around for the purposes of this PR?

It is possible that https://bitbucket.org/eigen/eigen/diff/unsupported/Eigen/src/AutoDiff/AutoDiffScalar.h?diff2=815734e048fd&at=default was what used to make this work?  It was removed around a year ago.

My best guess at a fix that doesn't involve putting `T{}` around every sub-expression would be to further specialize `make_coherent_internal` to handle Cwise ops, and/or possibly restore those removed Cwise specializations that forced MatrixXd intermediate results.  Some `drake/common` unit tests for all of this would also be in order.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 20 of 21 files reviewed at latest revision, 19 unresolved discussions.

---

*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-Ksq6BB-N0D-oX37frbe:b-8ljy9a) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Great.

The problem is in
```c++
    operator*(const AutoDiffScalar<OtherDerType>& other) const
    {
      internal::make_coherent(m_derivatives, other.derivatives());
      return MakeAutoDiffScalar(
        m_value * other.value(),
        (m_derivatives * other.value()) + (other.derivatives() * m_value));
    }
```
on the `(m_derivatives * other.value()) + (other.derivatives() * m_value));` part, with the failing assertion being `Assertion 'aLhs.rows() == aRhs.rows() && aLhs.cols() == aRhs.cols()' failed`.

The type of that expression is
```c++
Eigen::CwiseBinaryOp<
  Eigen::internal::scalar_sum_op<double, double>,
  const Eigen::CwiseBinaryOp<
    Eigen::internal::scalar_product_op<double, double>,
    const Eigen::CwiseBinaryOp<
      Eigen::internal::scalar_sum_op<double, double>,
      const Eigen::CwiseBinaryOp<
        Eigen::internal::scalar_product_op<double, double>,
        const Eigen::VectorXd,
        const Eigen::CwiseNullaryOp<
          Eigen::internal::scalar_constant_op<double>,
          const Eigen::VectorXd
        >
      >,
      const Eigen::CwiseBinaryOp<
        Eigen::internal::scalar_product_op<double, double>,
        const Eigen::VectorXd,
        const Eigen::CwiseNullaryOp<
          Eigen::internal::scalar_constant_op<double>,
          const Eigen::VectorXd
        >
      >
    >,
    const Eigen::CwiseNullaryOp<
      Eigen::internal::scalar_constant_op<double>,
      const Eigen::VectorXd
    >
  >,
  const Eigen::CwiseBinaryOp<
    Eigen::internal::scalar_product_op<double, double>,
    const Eigen::VectorXd,
    const Eigen::CwiseNullaryOp<
      Eigen::internal::scalar_constant_op<double>,
      const Eigen::VectorXd
    >
  >
>::CwiseBinaryOp(const Lhs &, const Rhs &, const BinaryOp &)
```

The coherency promotion signature is:
```c++
struct make_coherent_impl<Matrix<A_Scalar, A_Rows, A_Cols, A_Options, A_MaxRows, A_MaxCols>,
                             Matrix<B_Scalar, B_Rows, B_Cols, B_Options, B_MaxRows, B_MaxCols> > {
```

My guess is that the Cwise intermediate results are not matching the `Matrix` specialization of `make_coherent_impl`, but rather hitting the no-op default implementation of it.

If you force any of the intermediate multiplies back into a `T`, instead of a promise, I think it should stop failing, e.g.,
```c++
T test = T{params.mass() * params.length()} * state.theta();
```
This might be the best work-around for the purposes of this PR?

It is possible that https://bitbucket.org/eigen/eigen/diff/unsupported/Eigen/src/AutoDiff/AutoDiffScalar.h?diff2=815734e048fd&at=default was what used to make this work?  It was removed around a year ago.

My best guess at a fix that doesn't involve putting `T{}` around every sub-expression would be to further specialize `make_coherent_internal` to handle Cwise ops, and/or possibly restore those removed Cwise specializations that forced MatrixXd intermediate results.  Some `drake/common` unit tests for all of this would also be in order.
</blockquote></details>

Two quick thoughts:
1) good news is that it did abort in debug mode.  CI should be catching any offenses.
2) i think we agree that these problems should go away if/when we switch to fixed-size autodiff vector (instead of Xd)?  

my conclusion is that we upvote and prioritize #2619.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-Ksq7JHchWszexHGJA-V:bkgdsho) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
> i think we agree that these problems should go away if/when we switch to fixed-size autodiff vector

Yes, both from code reading, and also I just changed my spike test to try `AutoDiffd<8>` and it did not abort.

If `Xd` is going away entirely, then #2619 might suit.  If we think we might keep it for convenience anyway, or for long enough to keep stepping in this pothole too often, then I suspect a day or two by a C++ expert on our team would be enough to find the right specializations to make Xd work (probably by reaching out to the Eigen team, too; this is a pretty glaring bug IMO).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 20 of 21 files reviewed at latest revision, 19 unresolved discussions.

---

*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-Kss0pCyeD-O1TqQ3Lec:b-o5lw6c) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

> i think we agree that these problems should go away if/when we switch to fixed-size autodiff vector

Yes, both from code reading, and also I just changed my spike test to try `AutoDiffd<8>` and it did not abort.

If `Xd` is going away entirely, then #2619 might suit.  If we think we might keep it for convenience anyway, or for long enough to keep stepping in this pothole too often, then I suspect a day or two by a C++ expert on our team would be enough to find the right specializations to make Xd work (probably by reaching out to the Eigen team, too; this is a pretty glaring bug IMO).
</blockquote></details>

Would you post a github issue on drake, and perhaps also on eigen, summarizing your findings?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 20 of 21 files reviewed at latest revision, 19 unresolved discussions.

---

*[drake/examples/pendulum/energy_shaping_simulation.cc, line 42 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KsphNQZGLkYjX-rUdOW:-Kss3741hOVa2SgfrfFv:bmfxq1z) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/energy_shaping_simulation.cc#L42)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Wow, it sucks that `const auto& state = *(this->EvalVectorInput<PendulumState>(context, 0));` doesn't do what we want.  I wonder why it is documented to return nullptr rather than throw or abort?

In any case, consider `const auto* state = this->EvalVectorInput<PendulumState>(context, 0);` which seems like it should work, and pushes the cast gunk out of our way here.
</blockquote></details>

Done.  

---

*[drake/examples/pendulum/energy_shaping_simulation.cc, line 52 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Kspi9n2z7vY6x-3QWxx:-Kss69cT8657_-OBjCaW:bjmb3kc) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/energy_shaping_simulation.cc#L52)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI If you do `const T Etilde = // BR` to force a linebreak, you'll get ~14 more characters to work with, and I suspect it would linewrap more readably.
</blockquote></details>

cleaned up a little more (was planning to do it anyhow)

---

*[drake/examples/pendulum/energy_shaping_simulation.cc, line 53 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Kspi3aaqip6wbcUIjVD:-Kss6E-T_yZfRPJLMF9K:b-rh72dj) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/energy_shaping_simulation.cc#L53)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Why sometimes `pow(foo, 2)` and sometimes `foo * foo`?
</blockquote></details>

i prefer pow(,2), but haven't finished my cleanup pass of the code yet.  this is the first PR in a series.

---

*[drake/examples/pendulum/energy_shaping_simulation.cc, line 73 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Ksph-falbBxxPQmP9C_:-Kss7zbvHlki9M_R9eMv:b-1vi1ud) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/energy_shaping_simulation.cc#L73)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Should there be a TODO that the controller and plant somehow share or correspond their params?
</blockquote></details>

I like it better that they do not.  (I never have the true parameters of a real pendulum, but rather use estimated parameters in my controller).

---

*[drake/examples/pendulum/energy_shaping_simulation.cc, line 106 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspgQns2TYF5OaVopcp:-Kss8HFzQndBp8-ZbeRs:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/energy_shaping_simulation.cc#L106)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI This took me a little while to parse.  My final answer was "the discrepancy between our energy and the desired energy has at least halved during the simulation", which might be more clearly written as:
```
DRAKE_DEMAND(
    std::abs(final_energy - desired_energy) <
    (0.5 * std::abs(initial_energy - desired_energy)));
```
</blockquote></details>

Done.

---

*[drake/examples/pendulum/energy_shaping_simulation.cc, line 116 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Kspfe_GGH5RqygImh8V:-Kss8Lkm4g2FCHbsm7uI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/energy_shaping_simulation.cc#L116)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This needs `gflags::ParseCommandLineFlags(&argc, &argv, true);` in order for the command-line parameters to take effect.
</blockquote></details>

Done.

---

*[drake/examples/pendulum/lqr_simulation.cc, line 99 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspeW-ofBs0QgzxMV6Z:-KssA3m3IGbhTd2OOTfk:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/lqr_simulation.cc#L99)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This needs `gflags::ParseCommandLineFlags(&argc, &argv, true);` in order for the command-line parameters to take effect.
</blockquote></details>

Done.

---

*[drake/examples/pendulum/passive_simulation.cc, line 71 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspeaWYAMFQccbY1M5E:-KssA4budUaHvwoU3Uri:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/passive_simulation.cc#L71)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This needs `gflags::ParseCommandLineFlags(&argc, &argv, true);` in order for the command-line parameters to take effect.
</blockquote></details>

Done.

---

*[drake/examples/pendulum/pendulum_params.named_vector, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspdXXzI8T785UuZq7n:-KssAAEPcZFdLIB9cY_R:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/pendulum_params.named_vector#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unlikely ;)
</blockquote></details>

Done.

---

*[drake/examples/pendulum/pendulum_plant.cc, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Kspc4QD_Dp7Kh2aRMmw:-KssAMQRt6GCEz41BX3Z:bw2mo43) ([raw file](https://github.com/robotlocomotion/drake/blob/0ec3eba238258559c341a031564e7d2d0c2928e6/drake/examples/pendulum/pendulum_plant.cc#L20)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Why lose this check?
</blockquote></details>

i thought it was ugly and added nothing?

---

*[drake/examples/pendulum/pendulum_plant.cc, line 47 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspcpmIRIi2x_mcG2pa:-KssAUcnaTGNBaZfX5Uq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/pendulum_plant.cc#L47)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Needs `#include <cmath>`.
</blockquote></details>

Done.

---

*[drake/examples/pendulum/pendulum_plant.cc, line 51 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspczyDImDIWqBggZjm:-KssAskpnfyj0OTOu_n0:b-tm3c8t) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/pendulum_plant.cc#L51)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Both github and reviewable are rendering the unicode dot in thetadot atop the squared, not atop the theta (I had to zoom _way_ in to find it).  Is there some better unicode to use?
</blockquote></details>

let's see if this is better?

---

*[drake/examples/pendulum/pendulum_plant.cc, line 58 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspfK6plvvRabHh8n65:-KssAwGrFZ_DmEa5D_-M:b-iej8iv) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/pendulum_plant.cc#L58)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This is a beautiful implementation, but probably should have a (tiny) unit test to check it?
</blockquote></details>

i only added it to add tests to the examples (sigh...), but you're right.  done.

---

*[drake/systems/primitives/BUILD, line 324 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Kspj8QWuHlIWPqHxnFb:-KssD5w88nC3ObDFCyMp:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/BUILD#L324)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

CI is unhappy because `primitives` still has `CMakeLists.txt` but it doesn't mention pendulum (nor car it, because there's no CMake there).  I would say just de-list this unit test from CMake.
</blockquote></details>

Done.

---

*[drake/systems/primitives/linear_system.cc, line 102 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspWjwxV5I9jpAYLkuw:-KssD8ZArvf5egCvgupJ:brrup5r) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5c761b320f8ed0f5f38a18ba2df0e20a9c99dc/drake/systems/primitives/linear_system.cc#L102)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI I found this difficult to follow.  I think using a `std::tie` would be more clear.  Off the top of my head, that would look like:
```
AutoDiffXd autodiff_x;
AutoDiffXd autodiff_u;
std::tie(autodiff_x, autodiff_u) = math::initializeAutoDiffTuple(x0, u0);
autodiff_context->get_mutable_continuous_state_vector()->SetFromVector(
    autodiff_x);
if (num_inputs > 0) {
  autodiff_context->FixInputPort(0, autodiff_u);
}
```
</blockquote></details>

this code is sprinkled through a number of places (e.g. Linearize, trajectory optimization, ...).  we're due to cleanup the autodiff wrappers (or at least add sugar for working with autodiff + Context).   Future work.

---

*[drake/systems/primitives/linear_system.cc, line 116 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspWVDFkK4cmRztxm6v:-KssDo6QydLTMm_1mmgk:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5c761b320f8ed0f5f38a18ba2df0e20a9c99dc/drake/systems/primitives/linear_system.cc#L116)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Would this be clearer as `int num_gradients = std::get<0>(autodiff_args).derivs().size();` (or `autodiff_x.derivs.size()` with the new local variable suggestion, above), since I think the point is to match the other derivative vectors?
</blockquote></details>

Done.

---

*[drake/systems/primitives/linear_system.cc, line 124 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Ksp_F1K-taIy5ipLVaH:-KssDx3C406JUJlHg8vi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L124)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I can't figure out when this would end up not true -- perhaps if some system's AllocateParameters did a special-case override for AutoDiffXd that structured the derivs already somehow?  Probably a comment here is in order explaining why we have this guard.

Is there unit test coverage of the else condition (that size was non-zero, and was left unchanged)?
</blockquote></details>

Done.

---

*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-KssDyHppYO8W_PpZ_qv:b-tnervy) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Would you post a github issue on drake, and perhaps also on eigen, summarizing your findings?
</blockquote></details>

For this immediate PR, I think it's more robust to leave this initialization in the code right now, rather than push a strange fix into this system.  We have the same "fix" in trajectory optimization when it came up for simple car.  I think others are likely to hit it and be confused by it, as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Somehow 109174c got cherry-picked into this PR by accident.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-:-KssN6ZL813egbsf5OwC:bri0jp4)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 17 files at r1, 7 of 14 files at r4.
Review status: 22 of 29 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/pendulum/energy_shaping_simulation.cc, line 73 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Ksph-falbBxxPQmP9C_:-KssNgHMYpmObZwlrNMH:b-1wuldq) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/energy_shaping_simulation.cc#L73)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

I like it better that they do not.  (I never have the true parameters of a real pendulum, but rather use estimated parameters in my controller).
</blockquote></details>

OK That makes a lot of sense.  I'm not sure if it should appear in a comment somewhere, but please consider if so.

---

*[drake/examples/pendulum/pendulum_plant.cc, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Kspc4QD_Dp7Kh2aRMmw:-KssS4omoowuuwn9N5hK:beimgxq) ([raw file](https://github.com/robotlocomotion/drake/blob/0ec3eba238258559c341a031564e7d2d0c2928e6/drake/examples/pendulum/pendulum_plant.cc#L20)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

i thought it was ugly and added nothing?
</blockquote></details>

The intent was a cross-check that the `1, 1, 0` above precisely covered the codegen'd state size.  I guess that was unclear, though.

---

*[drake/examples/pendulum/pendulum_plant.cc, line 51 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspczyDImDIWqBggZjm:-KssNwhxieil50GchPfE:bl7z3ne) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/pendulum_plant.cc#L51)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

let's see if this is better?
</blockquote></details>

OK I can see it now.

---

*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-KssOFyiAYbVG83zSoE8:bq5umvn) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
Filed as #6944.  I will post more details there later on.

> I think it's more robust to leave this initialization in the code right now ...

Sure, but then this code needs a big TODO that it's a hack that should be ripped out later.  (Possibly also TODO the other places you mention, too.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 22 of 29 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/examples/pendulum/pendulum_plant.cc, line 18 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspbjNWp_UUoABqupS6:-Kt3lWbIdIXmzqhmNYtm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/examples/pendulum/pendulum_plant.cc#L18)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI We probably could (and should) promote the q/v/z count into the vector class's code-gen, so that they can be omitted here (or at worst, cited as named constants).  (Not for this PR.)
</blockquote></details>

Done.

---

*[drake/examples/pendulum/pendulum_plant.cc, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-Kspc4QD_Dp7Kh2aRMmw:-Kt3lIMUMVeLyNmVnZx0:b-94tksc) ([raw file](https://github.com/robotlocomotion/drake/blob/0ec3eba238258559c341a031564e7d2d0c2928e6/drake/examples/pendulum/pendulum_plant.cc#L20)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The intent was a cross-check that the `1, 1, 0` above precisely covered the codegen'd state size.  I guess that was unclear, though.
</blockquote></details>

DeclareContinuousState already checks the sum. 

---

*[drake/systems/primitives/linear_system.cc, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-KspYWwcJ_dKaEK0Q0Xt:-Kt3m5RuZCdcfwrgdFTh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ccde9c8fcb0e60dcd98ee6d716f75f25dc378b6a/drake/systems/primitives/linear_system.cc#L129)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Filed as #6944.  I will post more details there later on.

> I think it's more robust to leave this initialization in the code right now ...

Sure, but then this code needs a big TODO that it's a hack that should be ripped out later.  (Possibly also TODO the other places you mention, too.)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-:-Kt7YpQQO_38TFzjONgd:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
+@soonho-tri for platform review, please.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-:-Kt7dps5xXs7K8pnDece:bxwr5c8)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

FYI, I will clean-up the #6944-related hacks with my PR.

---

Reviewed 11 of 17 files at r1, 6 of 7 files at r2, 1 of 1 files at r3, 5 of 14 files at r4, 5 of 5 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6940#-:-KtIDbZHUppqKXKmk5aI:brocceu)*
<!-- Sent from Reviewable.io -->
